pennylane/transforms/decompositions/utils.py

```diff
@@ -0,0 +1,251 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions required for decomposing two-qubit operations."""
+
+from pennylane import math
+from pennylane import numpy as np
+
+# This gate E is called the "magic basis". It can be used to convert between
+# SO(4) and SU(2) x SU(2). For A in SO(4), E A E^\dag is in SU(2) x SU(2).
+E = np.array([[1, 1j, 0, 0], [0, 0, 1j, 1], [0, 0, 1j, -1], [1, -1j, 0, 0]]) / np.sqrt(2)
+Edag = E.conj().T
+
+LAST_COL_NEG = np.diag([1, 1, 1, -1])  # used to negate the last column of a matrix
+
+
+def _convert_to_su4(U):
+    r"""Check unitarity of a 4x4 matrix and convert it to :math:`SU(4)` if the determinant is not 1.
+
+    Args:
+        U (array[complex]): A matrix, presumed to be :math:`4 \times 4` and unitary.
+
+    Returns:
+        array[complex]: A :math:`4 \times 4` matrix in :math:`SU(4)` that is
+        equivalent to U up to a global phase.
+    """
+    # Check unitarity
+    if not math.allclose(math.dot(U, math.T(math.conj(U))), math.eye(4), atol=1e-7):
+        raise ValueError("Operator must be unitary.")
+
+    # Compute the determinant
+    det = math.linalg.det(U)
+
+    # Convert to SU(4) if it's not close to 1
+    if not math.allclose(det, 1.0):
+        exp_angle = -1j * math.cast_like(math.angle(det), 1j) / 4
+        U = math.cast_like(U, det) * math.exp(exp_angle)
+
+    return U
+
+
+def _compute_num_cnots(U):
+    r"""Compute the number of CNOTs required to implement a U in SU(4). This is based on
+    the trace of
+
+    .. math::
+
+        \gamma(U) = (E^\dag U E) (E^\dag U E)^T,
+
+    and follows the arguments of this paper: https://arxiv.org/abs/quant-ph/0308045.
+    """
+    u = math.dot(Edag, math.dot(U, E))
+    gammaU = math.dot(u, math.T(u))
+
+    trace = math.trace(gammaU)
+    evs = math.linalg.eigvals(gammaU)
+
+    # For the case with 3 CNOTs, the trace is a non-zero complex number
+    # with both real and imaginary parts.
+    num_cnots = 3
+
+    # At one point, to distinguish between 1/2 CNOT cases, we need to look at the eigenvalues
+    if math.get_interface(u) == "torch":
+        sorted_evs, _ = math.sort(math.imag(evs))
+    else:
+        sorted_evs = math.sort(math.imag(evs))
+
+    # Case: 0 CNOTs (tensor product), the trace is +/- 4
+    # We need a tolerance of around 1e-7 here in order to work with the case where U
+    # is specified with 8 decimal places.
+    if math.allclose(trace, 4, atol=1e-7) or math.allclose(trace, -4, atol=1e-7):
+        num_cnots = 0
```

Comment:
Does it make more sense to order it like this?

```python
trace = math.trace(gammaU)

if math.allclose(trace, 4, atol=1e-7) or math.allclose(trace, -4, atol=1e-7):
    return 0

evs = math.linalg.eigvals(gammaU)

if math.get_interface(u) == "torch":
    sorted_evs, _ = math.sort(math.imag(evs))
else:
     sorted_evs = math.sort(math.imag(evs))

if math.allclose(trace, 0j, atol=1e-7) and math.allclose(sorted_evs, [-1, -1, 1, 1]):
    return 1

if math.allclose(math.imag(trace), 0.0, atol=1e-7):
    return 2

return 3
```

? This way, you are avoiding computing EVs until they are needed.

Reply:
I'm curious if you can also swap the 1 and 2 CNOT case, to avoid sorting the evs unless needed, although it could be that the failure of the 1-CNOT case is required to verify the 2-CNOT case?

Reply:
The failure in the 1-CNOT case is indeed how I caught this, as the 2-CNOT cases were being sent to the 1-CNOT function. Either way, yes I can try to rearrange things :+1: